src/libsystemd/sd-event/sd-event.c

```diff
@@ -3931,8 +3970,71 @@ _public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
                                 assert_not_reached("Invalid wake-up pointer");
                         }
                 }
+                if (r < 0)
+                        return r;
+                if (r > 0)
+                        something_new = true;
+        }
+
+        *ret_min_priority = min_priority;
+        return something_new;
+}
+
+_public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
+        int r;
+
+        assert_return(e, -EINVAL);
+        assert_return(e = event_resolve(e), -ENOPKG);
+        assert_return(!event_pid_changed(e), -ECHILD);
+        assert_return(e->state != SD_EVENT_FINISHED, -ESTALE);
+        assert_return(e->state == SD_EVENT_ARMED, -EBUSY);
+
+        if (e->exit_requested) {
+                e->state = SD_EVENT_PENDING;
+                return 1;
+        }
+
+        for (int64_t threshold = INT64_MAX; ; threshold--) {
+                int64_t epoll_min_priority, child_min_priority, inotify_min_priority;
+
+                /* There may be a possibility that new epoll (especially IO) and child events are
+                 * triggered just after process_epoll() call but before process_child(), and the new IO
+                 * events may have higher priority than the child events. To salvage these events,
+                 * let's call epoll_wait() again, but accepts only events with higher priority than the
+                 * previous. See issue https://github.com/systemd/systemd/issues/18190 and comments
+                 * https://github.com/systemd/systemd/pull/18750#issuecomment-785801085
+                 * https://github.com/systemd/systemd/pull/18922#issuecomment-792825226 */
+
+                r = process_epoll(e, timeout, threshold, &epoll_min_priority);
+                if (r == -EINTR) {
+                        e->state = SD_EVENT_PENDING;
+                        return 1;
+                }
+                if (r < 0)
+                        goto finish;
+                if (r == 0 && threshold < INT64_MAX)
+                        /* No new epoll event. */
+                        break;
+
+                r = process_child(e, threshold, &child_min_priority);
+                if (r < 0)
+                        goto finish;
+                if (r == 0 && threshold < INT64_MAX)
+                        /* No new child event. */
+                        break;
+
+                r = process_inotify(e, threshold, &inotify_min_priority);
```

Comment:
hmm, it seems unnecessary to move process_inotify() into the loop. I.e. we have a separate inotify queue for each priority, hence there's no need to check the contents of each queue to determine the min prioriy on it, it's entirely enough to know that there's something inside the queue. This means we don't actually need to call process_inotify() inside the loop here, but it's enough to keep track of the priority where we ar about to call event_inotify_data_read() above, because at that point we already know the priority of what is queued.

Reply:
Oh, `struct inotify_data` has own priority field. I did not know that. Thanks.